packages/lib/bookings/filterHostsByLeadThreshold.ts

```diff
@@ -0,0 +1,135 @@
+import prisma from "@calcom/prisma";
+
+export const errorCodes = {
+  MAX_LEAD_THRESHOLD_FALSY: "Max lead threshold should be null or > 1, not 0.",
+} as const;
+
+function getMostRecentDate(dates: Date[]): Date {
+  if (dates.length === 0)
+    throw new Error("Array of date length provided in getMostRecentDate should not be empty.");
+
+  return dates.reduce((mostRecent, current) => (current > mostRecent ? current : mostRecent));
+}
+
+const computeLeadOffsets = async <T = Record<string, unknown>>({
+  hosts,
+  eventTypeId,
+}: {
+  eventTypeId: number;
+  hosts: (T & {
+    isFixed: false;
+    createdAt: Date;
+    user: {
+      id: number;
+    };
+  })[];
+}) => {
+  if (!hosts.length) return [];
+
+  const mostRecentDate = getMostRecentDate(hosts.map(({ createdAt }) => createdAt));
+  const userIds = hosts.map((host) => host.user.id);
+  // we need booking data now, this cannot be queried ahead of time as it requires knowing the most recent host date
+  // data only available after the initial call.
+  const bookingCounts = await prisma.booking.groupBy({
+    by: ["userId"],
+    where: {
```

Comment:
We need to fetch the same bookings here as we do in `getAllBookingsForRoundRobin`: no no-shows and cancellations, check for attendee emails

Reply:
This is done, moved the bookingCounts to the repository and created a builder for the where clause - it now shares the same where as getAllBookingsForRoundRobin